jac-uk/admin
# #2098 Vuexfire replacement

## What's included?
Closes #2098 

### ~~Solution 1~~
~~Replace `vuexfire` with [@xquick-code/vuexfire](https://www.npmjs.com/package/@xquick-code/vuexfire) in the short term fix.~~

Update:
1. The table pagination does not work well.

    https://github.com/jac-uk/admin/assets/79906532/1b585100-6957-48f9-a7f8-41bd33589cf5

### Solution 2
Writer our own helper method using `onSnapshot`.

The following changes need to be made in each store:

1. Change `import { firestoreAction } from 'vuexfire';` to `import { firestoreAction } from '@/helpers/vuexfireJAC';`.
2. Add a local mutation to each store.

    ```js
    mutations: {
        set(state, { name, value }) {
            state[name] = value;
        },
    }
    ```

    Note: I tried to use the global mutations in the `firestoreAction` but I couldn't find an efficient way to get the namespace of each store.


## Who should test?
✅ Product owner
✅ Developers

## How to test?
Go to each page on admin and check if the data is reactive. 

For example:
- Find an exercise. (e.g. https://jac-admin-develop--pr2102-feature-2098-vuexfir-m382fggx.web.app/exercise/haDRCCKfsxsL7gCWGy3X/details/overview)
- Change the exercise name manually in Firestore
- Check if the UI updates automatically without refreshing the page
- Archive/unarchive/delete an exercise and check if the UI updates automatically without refreshing the page

## Risk - how likely is this to impact other areas?
🔴 High risk - this includes a lot of changes to shared code

## Additional context
Include screen grabs, video demo, notes etc.

## Related permissions
Have permissions been considered for this functionality?
- No permission changes required
- Permissions have been added / updated. Details:

---
PREVIEW:DEVELOP
_can be OFF, DEVELOP or STAGING_
